### PR TITLE
Run cron jobs in Docker

### LINKED ISSUE #4383: Create cron jobs docker-compose service
~Fixed in #4387~
Fixed in ~internetarchive/olsystem#118~
Subtask of: #4060

We have a number of cron jobs that are run on `ol-home` ; in order to run this in our docker-based `ol-home0` node, we should place them in a service. Note these cron jobs live in olsystem.

* [x] Transfer ol-home cron jobs to a new cron file.
* [x] Ensure all ol-home cron jobs run on ol-home0
    * [x]  #4622 `/olsystem/bin/cron/oldump.sh ol-db1` --> ~#4627~ + internetarchive/olsystem#118
        * `dumping the data table ` # for a long time... 102 minutes!
        * Fixed in this PR by correcting `$SCRIPTS` in `bin/cron/oldump.sh`
        * Got an error on a job on the IA side; it was trying to rsync from the docker container: `rsync://e1d610933819/var_1/tmp/dumps/ol_cdump_2021-01-30`
        * STATE: Running; waiting to see if it completes
    * [x] `$SCRIPTS/store_counts.py` * 2 -- fixed in internetarchive/infogami#134
        * [x] `if _is_illegal_header_value(values[i]):` TypeError: expected string or bytes-like object 
        * [x] Errored: Unable to access infobase?
    * [x] `$SCRIPTS/2011/08/expire_accounts.py` -- fixed in internetarchive/infogami#134
        * [x] `if _is_illegal_header_value(values[i]):` TypeError: expected string or bytes-like object 
        * [x] Errored: Unable to access infobase?
              * Needed to change infobase link from ol-home0 -> ol-home in openlibrary.yml
    * [x] `openlibrary.solr.process_stats` * 2 (loan stats)
        * `cat /var/log/openlibrary/loanstats.log`
        * [x] `/run/secrets/ia_db_pw_file: not found` should have been fixed by #115
        * From https://github.com/internetarchive/olsystem/blob/master/etc/openlibrary.yml#L31
        * [x] #4542 Python 3 decode error for 1 loan?
    * [x] `python $SCRIPTS/manage-imports.py`
        * [x] `/run/secrets/ia_db_pw_file: not found` should have been fixed by #115
        * No errors now.
* [x] Ensure all ol-www1 cron jobs run in docker + py3
    * [x] (Moved to #4775 and separated from this issue) sitemap rsync should run on ol-home0 [Slack thread](https://internetarchive.slack.com/archives/C075DJ6G1/p1614111981011600) <details>The sitemaps are generated on ol-home0 and copied over to ol-www1 where they are served. We can actually make the `rsync` command run on ol-home0 and target ol-www1 (and/or ol-www0). This is **not** blocked by ol-www0 migration.  `saddr $CLUSTER proto tcp dport rsync ACCEPT;` in `/etc/ferm/ferm.conf` is vital.</details>
    * [x] cron-jobs service stuck in restart loop #4707 fixed in #4717
    * [x] (Moved to #4775 and separated from this issue) MOVE ALL CRON JOBS TO OL-HOME0 ipstats.py. Unclear what ideal course of action is for these, but if we could get them running on ol-home0 with other cron jobs: that would be nice!

### PR DESCRIPTION
**Pull request by cclauss**
Blocked by: ~internetarchive/olsystem#110~
Closes #4383, #4542
Subtask of #4060

We are running on an oldev-based node but we want to run the cronjobs that are in [`/olsystem/etc/cron.d`](https://github.com/internetarchive/olsystem/blob/master/etc/cron.d).

What about `Caution on the double run`?

https://stackoverflow.com/questions/37458287/how-to-run-a-cron-job-inside-a-docker-container

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -153,6 +153,26 @@ services:
       options:
         max-size: "512m"
         max-file: "4"
+  cronjobs:
+    image: oldev:latest
+    build:
+      context: .
+      dockerfile: docker/Dockerfile.oldev
+    command: docker/ol-cronjobs-start.sh
```
Comment (cclauss) [inline docker-compose.yml:161]: Do we need `COPY /olsystem/etc/cron.d /etc/cron.d/`

File: docker-compose.yml [hunk 2]
```diff
@@ -153,6 +153,26 @@ services:
...
+    restart: always
+    hostname: "$HOSTNAME"
```
Comment (cclauss) [inline docker-compose.yml:163]: Remove hostname?

File: docker-compose.yml [hunk 3]
```diff
@@ -153,6 +153,26 @@ services:
...
+    environment:
```
Comment (cclauss) [inline docker-compose.yml:164]: Remove environment?

File: docker-compose.yml [hunk 4]
```diff
@@ -153,6 +153,26 @@ services:
...
+      - OL_CONFIG=conf/openlibrary.yml
+      - PYENV_VERSION=${PYENV_VERSION:-}
+    expose:
```
Comment (cclauss) [inline docker-compose.yml:167]: Remove expose?

File: docker-compose.yml [hunk 5]
```diff
@@ -153,6 +153,26 @@ services:
...
+      - 7700
+    depends_on:
```
Comment (cclauss) [inline docker-compose.yml:169]: Remove depends_on?

File: docker-compose.yml [hunk 6]
```diff
@@ -153,6 +153,26 @@ services:
...
+      - db
+    volumes:
+      - .:/openlibrary
+    networks:
+      - webnet
+      - dbnet
```
Comment (cclauss) [inline docker-compose.yml:175]: Remove dbnet?

File: docker-compose.yml (hunk 1 above)
Comment (cclauss) [reply docker-compose.yml:161]: This file: openlibrary -> /olsystem/etc/cron.d/openlibrary.allnodes

**Comment by cclauss** [comment]
Add  ia_db_pw_file for 5 * * * * python $SCRIPTS/openlibrary-server /olsystem/etc/openlibrary.yml runmain openlibrary.solr.process_stats --load-from-db > /var/log/openlibrary/loanstats.log 2>&1

File: docker-compose.production.yml [hunk 7]
```diff
@@ -55,6 +55,22 @@ services:
       # Needed by default-docker.conf
       - ssl_certificate
       - ssl_certificate_key
+  cron-jobs:
+    image: oldev:latest
+    build:
+      context: .
+      dockerfile: docker/Dockerfile.oldev
+    command: cron -f
```
Comment (cdrini) [inline docker-compose.production.yml:63]: looks like this will need the `HOSTNAME` environment variable.

**Comment by cclauss** [comment]
Added the solrwriter change where both `strip_bad_char()` and [`unicodedata.normalize()`](https://docs.python.org/3/library/unicodedata.html#unicodedata.normalize) expect `value` to be a `str`.

**Comment by cclauss** [comment]
```html
<html>

<head>
    <title>Apache Tomcat/6.0.39 - Error report</title>
    <style>
        <!--H1 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:22px;} H2 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:16px;} H3 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:14px;} BODY {font-family:Tahoma,Arial,sans-serif;color:black;background-color:white;} B {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;} P {font-family:Tahoma,Arial,sans-serif;background:white;color:black;font-size:12px;}A {color : black;}A.name {color : black;}HR {color : #525D76;}
        -->
    </style>
</head>

<body>
    <h1>HTTP Status 400 - ERROR: [doc=loans/loan-1944635007] Error adding field
        \'duration_hours_i\'=\'0.0008333333333333334\'</h1>
    <HR size="1" noshade="noshade">
    <p><b>type</b> Status report</p>
    <p><b>message</b> <u>ERROR: [doc=loans/loan-1944635007] Error adding field
            \'duration_hours_i\'=\'0.0008333333333333334\'</u></p>
    <p><b>description</b> <u>The request sent by the client was syntactically incorrect.</u></p>
    <HR size="1" noshade="noshade">
    <h3>Apache Tomcat/6.0.39</h3>
</body>

</html>
```

**Review by cdrini — CHANGES_REQUESTED** [review]
Lgtm! One last change: Can you update our deploy script to make sure it starts this service?

**Comment by cclauss** [comment]
I made that change on #4457 but that is still not landed.

**Review by mekarpeles — APPROVED** [review]
lgtm

**Comment by mekarpeles** [comment]
I see restart always on the cron and #4457 has landed, going to merge this as well.

**Pull request merged**